PDDL.py

```diff
@@ -95,9 +97,29 @@ def parse_predicates(self, group):
                 arguments[untyped_variables.pop(0)] = 'object'
             self.predicates[predicate_name] = arguments
 
-    #-----------------------------------------------
+    # -----------------------------------------------
+    # Parse types
+    # -----------------------------------------------
+    def parse_types(self, types):
+        untyped_names = []
+        while types:
+            t = types.pop(0)
+            if self.types.get(t):
+                raise Exception('Type ' + t + ' redefined')
+            elif t == '-':
+                if not untyped_names:
+                    raise Exception('Unexpected hyphen in types')
+                typedef = types.pop(0)
+                while untyped_names:
+                    self.types[typedef].append(untyped_names.pop(0))
+            else:
+                untyped_names.append(t)
+        while untyped_names:
+            self.types['object'].append(untyped_names.pop(0))
```

Comment:
I guess most people use object as the root supertype, not only for the untyped ones.

Reply:
I've been looking at some PDDL examples from Pyperplan for reference on this. As one instance, the TPP Propositional domain uses the following definition for :types ...
`(:types place locatable level - object
	depot market - place
	truck goods - locatable)`
I could be wrong (and often am!), but it seemed as though the only way to have leftover (i.e., unassigned) untyped names when the preceding loop ended would be if the names had `object` as the root supertype. All the other types would seem to have been assigned. Or am I missing what you were trying to tell me?

Reply:
Some domains contain the entire tree types explicitly defined, others may lack the ``object`` ones defined at the end.
I am not sure how legal/valid this PDDL would be, but it would cost even less to just ignore the untyped ones and just assume object instead.
I mean, what happens if we ignore this last while?

Reply:
I guess the question here would be, would assuming an object type break anything?

Reply:
Not worthy it I guess, some people even describe cycles in the typing hierarchy.
Even the IPC2020 did that, it took me a while to see it.
This is the reason why HyperTensioN's HDDL type parser is much stronger than the PDDL one.